### PR TITLE
Add outputPath configuration option

### PR DESCRIPTION
**Pull request by jklukas**
Implements the feature discussed in #12. 

@tkrille - Do you think this needs a test? File I/O is a generally ugly area to write unit tests for.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ First add the dependency to your POM:
 <dependency>
     <groupId>de.thomaskrille</groupId>
     <artifactId>dropwizard-template-config</artifactId>
-    <version>1.2.0</version>
+    <version>1.3.0</version>
```
Comment (tkrille) [inline README.md]: Please revert this change. I will do this on the release day.

File: README.md [hunk 2]
```diff
@@ -43,7 +43,9 @@ public void initialize(final Bootstrap<Configuration> bootstrap) {
 }
 ```
 
-Look at `TemplateConfigBundleConfiguration`'s javadoc to see all available options.
+Look at
+[`TemplateConfigBundleConfiguration`](src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigBundleConfiguration.java)'s
+javadoc to see all available options.
```
Comment (tkrille) [inline README.md]: Thanks for cleaning up the README, but please do it in a separate commit.

File: pom.xml [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
     <groupId>de.thomaskrille</groupId>
     <artifactId>dropwizard-template-config</artifactId>
-    <version>1.2.1-SNAPSHOT</version>
+    <version>1.3.1-SNAPSHOT</version>
```
Comment (tkrille) [inline pom.xml]: You can update to `1.3.0-SNAPSHOT`, but `1.3.1-SNAPSHOT` doesn't make any sense :)

**Comment by tkrille** [comment]
There should definitively be tests for this. I don't mind dealing with file I/O in tests :)

**Comment by tkrille** [comment]
It would be nice, if you write some documentation for this feature.

**Comment by tkrille** [comment]
With "documentation", I mean the tutorial part in the README. Just to be clear :)

**Comment by tkrille** [comment]
Thanks four your efforts. I really appreciate it. The code looks good to me. I'm looking forward to actually test this in the evening.

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigBundleConfiguration.java [hunk 4]
```diff
@@ -38,12 +39,27 @@ public TemplateConfigBundleConfiguration charset(Charset charset) {
     }
 
     /**
+     * Get the configured output path for the processed config (Default: None)
+     */
+    public Optional<String> outputPath() { return outputPath; }
```
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigBundleConfiguration.java]: Please, write this on multiple lines.

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java [hunk 5]
```diff
@@ -69,8 +75,15 @@ public InputStream open(final String path) throws IOException {
 
             new Template("config", configTemplate, configuration)
                     .process(dataModel, new OutputStreamWriter(processedTemplateStream, charset));
+            byte[] processedTemplateBytes = processedTemplateStream.toByteArray();
 
-            return new ByteArrayInputStream(processedTemplateStream.toByteArray());
+            if (outputPath.isPresent()) {
+                File outputFile = new File(outputPath.get());
+                Files.createParentDirs(outputFile);
+                Files.write(processedTemplateBytes, outputFile);
```
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java]: I'm not sure, but should some error handling be added or does Dropwizard fail "gracefully"?

Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java]: Or should errors simply be ignored (but logged)? In the end, writing the file is not crucial for the bundle to work.

Comment (jklukas) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java]: It's not entirely clear to me how Dropwizard will handle an error here. Catching the error and logging sounds like a reasonable option, but I don't think logging in an option here since this code is getting executed before the configuration is processed, so logging isn't set up yet and the messages won't be going anywhere.

File: README.md (hunk 1 above)
Comment (jklukas) [inline README.md]: Will revert.

**Comment by jklukas** [comment]
Just pushed a new commit to address all comments except for the one about error handling while writing the file. I'll take a look at what Dropwizard does when this code section throws an error.

**Comment by jklukas** [comment]
We can squash this down to one commit before merging if you'd like. Let me know your preference.

**Comment by jklukas** [comment]
As an example of how Dropwizard fails on an exception being thrown, here's a stacktrace I got when passing in `null` for outputPath:
```
Exception in thread "main" java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:210)
	at com.google.common.base.Optional.of(Optional.java:85)
	at de.thomaskrille.dropwizard_template_config.TemplateConfigBundleConfiguration.outputPath(TemplateConfigBundleConfiguration.java:64)
	at com.simple.estuary.demux.EstuaryDemuxApplication.initialize(EstuaryDemuxApplication.java:34)
	at io.dropwizard.Application.run(Application.java:75)
	at com.simple.estuary.demux.EstuaryDemuxApplication.main(EstuaryDemuxApplication.java:23)
```

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java (hunk 5 above)
Comment (jklukas) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java]: I'd also love to log an INFO-level message here to indicate that the rendered config has been written to X path, but again I don't think we can do any useful logging at this point in the lifecycle.

Comment (jklukas) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java]: If the user has configured the package to write out the config, it might be reasonable not to handle exceptions here and just let the exception abort startup.

**Comment by jklukas** [comment]
I also still need to add tests.

**Comment by jklukas** [comment]
Pushed a new commit adding a test that checks `outputPath` correctly creates a file.

**Comment by jklukas** [comment]
This should be good for more review now. Let me know if the test looks sufficient and if you have further opinions on exception handling / logging.

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java (hunk 5 above)
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigurationSourceProvider.java]: Alright, agreed. I didn't think about the order of logging setup and parsing the config.

**Comment by tkrille** [comment]
Thanks for your awesome work! I also thinks is good to merge now.

> We can squash this down to one commit before merging if you'd like. Let me know your preference.

Yes, please do this now and then I will merge this. I assume, that you want a release of 1.3.0 as soon as possible, so I will probably release it on Saturday.

**Comment by jklukas** [comment]
This is now squashed to a single commit. I personally don't have a need to rush out a 1.3.0 release to include this. Implementing https://github.com/tkrille/dropwizard-template-config/issues/16 is probably going to be necessary to make this project a compelling replacement for what my colleagues are currently doing.

**Comment by tkrille** [comment]
Thanks a lot and sorry for the delay.

**Comment by tkrille** [comment]
Ok, then let's implement #16 first.
